sql/core/src/test/scala/org/apache/spark/sql/QueryTest.scala

```diff
@@ -350,6 +350,9 @@ object QueryTest {
     Row.fromSeq(row.toSeq.map {
       case null => null
       case d: java.math.BigDecimal => BigDecimal(d)
+      // Round up the values to avoid approximate error
+      case f : Float => BigDecimal(f).setScale(8, BigDecimal.RoundingMode.HALF_UP).toFloat
```

Comment:
I have also tested this case: 
val a = 1e15
val b = 1e15 - 1
val c = BigDecimal(a).setScale(16, BigDecimal.RoundingMode.HALF_UP).toDouble
// c: Double = 1.0E15
val d = BigDecimal(b).setScale(16, BigDecimal.RoundingMode.HALF_UP).toDouble
// d: Double = 9.99999999999999E14
c == d
// res6: Boolean = false